nlp_utilities.jl

```diff
@@ -0,0 +1,423 @@
+    W[1:num_vars, 1:num_vars] = hessian[1:num_vars, 1:num_vars]
+    # Jacobian of the constraints wrt the primal variables
+    A = spzeros(num_cons, num_vars + num_ineq)
+    A[:, 1:num_vars] = jacobian[:, 1:num_vars]
+    for (i,j) in enumerate(ineq_locations)
+        A[j, num_vars+i] = -1
+    end
+    # Partial second derivative of the lagrangian wrt primal solution and parameters
+    ∇ₓₚL = spzeros(num_vars + num_ineq, num_parms)
+    ∇ₓₚL[1:num_vars, :] = hessian[1:num_vars, num_vars+1:end]
+    # Partial derivative of the equality constraintswith wrt parameters
+    ∇ₚC = jacobian[:, num_vars+1:end]
+
+    # M matrix
+    # M = [
+    #     [W A' -I I];
+    #     [A 0 0 0];
+    #     [V_L 0 (X - X_L) 0]
+    #     [V_U 0 0 0 (X_U - X)]
+    # ]
+    len_w = num_vars + num_ineq
+    M = spzeros(len_w + num_cons + num_low + num_up, len_w + num_cons + num_low + num_up)
+
+    M[1:len_w, 1:len_w] = W
+    M[1:len_w, len_w + 1 : len_w + num_cons] = A'
+    M[len_w+1:len_w+num_cons, 1:len_w] = A
+    M[1:len_w, len_w+num_cons+1:len_w+num_cons+num_low] = I_L
+    M[len_w+num_cons+1:len_w+num_cons+num_low, 1:len_w] = V_L
+    M[len_w+num_cons+1:len_w+num_cons+num_low, len_w+num_cons+1:len_w+num_cons+num_low] = X_lb
+    M[len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up, 1:len_w] = V_U
+    M[len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up, len_w+num_cons+num_low+1:len_w+num_cons+num_low+num_up] = X_ub
+    M[1:len_w, len_w+num_cons+num_low+1:end] = I_U
+
+    # N matrix
+    # N = [∇ₓₚL ; ∇ₚC; zeros(num_low + num_up, num_parms)]
+    N = spzeros(len_w + num_cons + num_low + num_up, num_parms)
+    N[1:len_w, :] = ∇ₓₚL
+    N[len_w+1:len_w+num_cons, :] = ∇ₚC
+
+    return M, N
+end
+
+"""
+    compute_derivatives_no_relax(evaluator::MOI.Nonlinear.Evaluator, cons::Vector{ConstraintRef},
+        primal_vars::Vector{VariableRef}, params::Vector{VariableRef},
+        _X::Vector, _V_L::Vector, _X_L::Vector, _V_U::Vector, _X_U::Vector, ineq_locations::Vector{Z},
+        has_up::Vector{Z}, has_low::Vector{Z}
+    ) where {Z<:Integer}
+
+Compute the derivatives of the solution w.r.t. the parameters without accounting for active set changes.
+"""
+function compute_derivatives_no_relax(evaluator::MOI.Nonlinear.Evaluator, cons::Vector{ConstraintRef},
+    primal_vars::Vector{VariableRef}, params::Vector{VariableRef}, 
+    _X::AbstractVector, _V_L::AbstractVector, _X_L::AbstractVector, _V_U::AbstractVector, _X_U::AbstractVector, ineq_locations::Vector{Z},
+    has_up::Vector{Z}, has_low::Vector{Z}
+) where {Z<:Integer}
+    num_bounds = length(has_up) + length(has_low)
+    M, N = build_M_N(evaluator, cons, primal_vars, params, _X, _V_L, _X_L, _V_U, _X_U, ineq_locations, has_up, has_low)
+
+    # Sesitivity of the solution (primal-dual_constraints-dual_bounds) w.r.t. the parameters
+    K = lu(M) # Factorization
+    ∂s = zeros(size(M, 1), size(N, 2))
+    # ∂s = - (K \ N) # Sensitivity
+    ldiv!(∂s, K, N)
```

Comment:
@frapac, I have passed the code to use sparse matrices :). Let me know if something could be simpler or more efficient.

The only limitations I hadn't foreseen so far are: 
-  the requirement of the output matrix (e.g., `∂s`) to be dense;
- the only factorization I managed to make things work is the `lu`. Perhaps this is obvious but was not something I anticipated.

Reply:
ps. tests continue to pass!

Reply:
thank you for the head up @andrewrosemberg !

> the only factorization I managed to make things work is the lu.

This is expected. The KKT matrix is symmetric indefinite so most sparse linear factorization won't work there. LU is an option. Another is to use a LBL factorization. 

Ipopt uses a LBL factorization internally in its algorithm. Problem is that there are only a few sparse linear solvers implementing the sparse LBL: MUMPS (open-source), Pardiso (close source) or HSL ma27/ma57 (the favored option in Ipopt). 

I would be in to use LU by default (as UMFPACK is shipped directly with Julia), and add an extension to compute the sensitivity with HSL if the user has access to it. 

To use HSL, HSL.jl is the recommended wrapper: https://github.com/JuliaSmoothOptimizers/HSL.jl 

In addition, the solver HSL ma57 supports solving a sparse linear system with multiple right hand side, so that could help in speeding up the evaluation of the sensitivity here.